internal/storage/memory.go

```diff
@@ -361,3 +387,230 @@ func inMemoryCRDB() (testserver.TestServer, error) {
 
 	return ts, nil
 }
+
+type memoryUserInfoService struct {
+	db         *sql.DB
+	httpClient *http.Client
+}
+
+type userInfoServiceOpt func(*memoryUserInfoService)
+
+func newUserInfoService(config Config, opts ...userInfoServiceOpt) (*memoryUserInfoService, error) {
+	s := &memoryUserInfoService{
+		db:         config.db,
+		httpClient: http.DefaultClient,
+	}
+
+	for _, opt := range opts {
+		opt(s)
+	}
+
+	err := s.createTables()
+
+	return s, err
+}
+
+// WithHTTPClient allows configuring the HTTP client used by
+// memoryUserInfoService to call out to userinfo endpoints.
+func WithHTTPClient(client *http.Client) func(svc *memoryUserInfoService) {
+	return func(svc *memoryUserInfoService) {
+		svc.httpClient = client
+	}
+}
+
+func (s *memoryUserInfoService) createTables() error {
+	_, err := s.db.Exec(`
+        CREATE TABLE IF NOT EXISTS user_info (
+            id    UUID PRIMARY KEY NOT NULL DEFAULT gen_random_uuid(),
+            name  STRING NOT NULL,
+            email STRING NOT NULL,
+            sub   STRING NOT NULL,
+            iss_id   UUID NOT NULL REFERENCES issuers(id),
+            UNIQUE (iss_id, sub)
+        )`)
+
+	return err
+}
+
+// LookupUserInfoByClaims fetches UserInfo from the store.
+// This does not make an HTTP call with the subject token, so for this
+// data to be available, the data must have already be fetched and
+// stored.
+func (s memoryUserInfoService) LookupUserInfoByClaims(ctx context.Context, iss, sub string) (*types.UserInfo, error) {
+	selectCols := withQualifier([]string{
+		userInfoCols.Name,
+		userInfoCols.Email,
+		userInfoCols.Subject,
+	}, "ui")
+
+	selectCols = append(selectCols, "i."+issuerCols.URI)
+
+	selects := strings.Join(selectCols, ",")
+
+	stmt := fmt.Sprintf(`
+	SELECT %[1]s FROM user_info ui
+        JOIN issuers i ON ui.%[2]s = i.%[3]s
+        WHERE i.%[4]s = $1 and ui.%[5]s = $2`,
+		selects,
+		userInfoCols.IssuerID,
+		issuerCols.ID,
+		issuerCols.URI,
+		userInfoCols.Subject,
+	)
+	row := s.db.QueryRowContext(ctx, stmt, iss, sub)
+
+	var ui types.UserInfo
+
+	err := row.Scan(&ui.Name, &ui.Email, &ui.Subject, &ui.Issuer)
+
+	if errors.Is(err, sql.ErrNoRows) {
+		return nil, types.ErrUserInfoNotFound
+	}
+
+	return &ui, err
+}
+
+func (s memoryUserInfoService) LookupUserInfoByID(ctx context.Context, id string) (*types.UserInfo, error) {
+	selectCols := withQualifier([]string{
+		userInfoCols.ID,
+		userInfoCols.Name,
+		userInfoCols.Email,
+		userInfoCols.Subject,
+	}, "ui")
+
+	selectCols = append(selectCols, "i."+issuerCols.URI)
+
+	selects := strings.Join(selectCols, ",")
+
+	stmt := fmt.Sprintf(`
+        SELECT %[1]s FROM user_info ui
+        JOIN issuers i ON ui.%[2]s = i.%[3]s
+        WHERE ui.%[4]s = $1
+        `, selects, userInfoCols.IssuerID, issuerCols.ID, userInfoCols.ID)
+
+	row := s.db.QueryRowContext(ctx, stmt, id)
```

Comment:
The [DB logic][db-logic] for getting issuers uses a transaction if one exists and the DB connection if it doesn't, since we don't strictly _need_ to begin transactions if we're just fetching data. I'd recommend doing the same thing here.

[db-logic]: https://github.com/infratographer/identity-manager-sts/blob/d77fa3b76ee80422cd4666a1d641c89ea0188d94/internal/storage/memory.go#L213-L220